### PR TITLE
Exposing raw HTTP response to REST API consumers

### PR DESCRIPTION
**Pull request by bhariharan**

### CODE REVIEW

**Comment by huminzhi** [comment]
## Danger has errored
[!] Invalid `DangerFile` file: `minimum_coverage_percentage` value must be a Float! Found Integer instead.
```
 #  from DangerFile:5
 #  -------------------------------------------
 #  
 >  xcov.report(
 #     scheme: 'UnitTests',
 #  -------------------------------------------
```

<p align="right" data-meta="generated_by_CI-result">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by khawkins** [comment]
Do we plan to update the blocks interfaces too?  That would be really helpful, as a lot of consumers utilize the blocks for success/failure management, rather than the delegates.

**Comment by bhariharan** [comment]
Yup - I'm working on it. Lot of code changes on the blocks ;-)

**Comment by bhariharan** [comment]
I got the blocks too. Running sample apps and tests now.

**Comment by bhariharan** [comment]
Tested `SmartSyncExplorer`, `RestAPIExplorer` and `SmartSyncExplorerHybrid` - all seem to be working fine.

**Comment by khawkins** [comment]
Seems good to me, overall.  I had a random problem with the unit tests in Xcode 9, which was fixed by changing `self.view` to `_view` in [this line of SFOAuthCoordinator.m](https://github.com/forcedotcom/SalesforceMobileSDK-iOS/blob/dev/libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFOAuthCoordinator.m#L284).  Do you want to add that update to your PR, or for me to send a separate PR for it?

Otherwise, this looks good to me!

**Comment by bhariharan** [comment]
@khawkins Fixed [here](https://github.com/forcedotcom/SalesforceMobileSDK-iOS/pull/2208/commits/c5001efee8c187507b1bbae6e4f02ebf8180a53e).

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestAPI.m [hunk 1]
```diff
@@ -382,16 +382,16 @@ - (void)sendPendingRequests {
     }
 }
 
-- (void)notifyDelegateOfResponse:(id<SFRestDelegate>)delegate request:(SFRestRequest *)request data:(id)data {
-    if ([delegate respondsToSelector:@selector(request:didLoadResponse:)]) {
-        [delegate request:request didLoadResponse:data];
+- (void)notifyDelegateOfResponse:(id<SFRestDelegate>)delegate request:(SFRestRequest *)request data:(id)data rawResponse:(NSURLResponse *)rawResponse {
+    if ([delegate respondsToSelector:@selector(request:didLoadResponse:rawResponse:)]) {
+        [delegate request:request didLoadResponse:data rawResponse:rawResponse];
```
Comment (asikora-sf) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestAPI.m]: @bhariharan @khawkins Can we get a fallback to the non-raw-response version of this? I.e.
`else if ([delegate respondsToSelector:@selector(request:didLoadResponse:)` to support backwards compatibility?

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestAPI.m [hunk 2]
```diff
@@ -382,16 +382,16 @@ - (void)sendPendingRequests {
...
     }
     [self removeActiveRequestObject:request];
 }
 
-- (void)notifyDelegateOfFailure:(id<SFRestDelegate>)delegate request:(SFRestRequest *)request error:(NSError *)error {
-    if ([delegate respondsToSelector:@selector(request:didFailLoadWithError:)]) {
-        [delegate request:request didFailLoadWithError:error];
+- (void)notifyDelegateOfFailure:(id<SFRestDelegate>)delegate request:(SFRestRequest *)request error:(NSError *)error rawResponse:(NSURLResponse *)rawResponse {
+    if ([delegate respondsToSelector:@selector(request:didFailLoadWithError:rawResponse:)]) {
+        [delegate request:request didFailLoadWithError:error rawResponse:rawResponse];
```
Comment (asikora-sf) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestAPI.m]: Same here

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestAPI.m (hunk 1 above)
Comment (bhariharan) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/RestAPI/SFRestAPI.m]: @asikora-sf Yeah, it's done [here](https://github.com/forcedotcom/SalesforceMobileSDK-iOS/pull/2210).

**Pull request merged**